[sully] mysql 연동 / soft-delete / 댓글 작성, 삭제 / 댓글 수 카운트 / ajax 댓글 작성

devbattery opened the pull request:
> 특히 ajax에서 벽을 많이 느꼈던 일주일이었습니다. 이번주도 잘 부탁드립니다 로치! 🙇🏻

<br>

- 현재 ajax로 수정한 상태라 기존의 댓글 작성과 삭제는 찍지 못 하는 상태입니다.
  - 미리 찍어놓았어야 했는데 미처 생각을 못 했습니다 ㅠㅠ
  - 이번 PR은 `결과 GIF`를 대신하여 코드를 중심으로 작성하겠습니다.

<br>

# 배포 주소

- http://43.201.10.160:8080/

<br>

# SQL

```sql
CREATE TABLE USERS
(
    id       BIGINT AUTO_INCREMENT NOT NULL,
    user_id  VARCHAR(13)           NOT NULL,
    password VARBINARY(32)         NOT NULL,
    name     VARCHAR(10)           NOT NULL,
    email    VARCHAR(30)           NOT NULL,
    PRIMARY KEY (id)
);

CREATE TABLE ARTICLES
(
    id          BIGINT AUTO_INCREMENT NOT NULL,
    user_id     VARCHAR(13)           NOT NULL,
    title       VARCHAR(30)           NOT NULL,
    contents    TEXT                  NOT NULL,
    currentTime TIMESTAMP             NOT NULL,
    deleted     BOOLEAN,
    PRIMARY KEY (id)
);

CREATE TABLE REPLIES
(
    id            BIGINT AUTO_INCREMENT NOT NULL,
    article_id    INT                   NOT NULL,
    user_id       VARCHAR(13)           NOT NULL,
    reply_content TEXT                  NOT NULL,
    reply_time    TIMESTAMP             NOT NULL,
    deleted       BOOLEAN,
    PRIMARY KEY (id)
);
```

<br>

# 구현 내용 (많이 고민했던 부분만)

## mysql 연동

- cs16 기간 중 docker에 mysql을 설치하여 진행한 적이 있어서 로컬에서는 금방 h2에서 mysql로 변경할 수 있었습니다.
- 로컬은 큰 문제가 없었지만 aws 연동에서 많은 시간을 소비했습니다.
  - 그래도 호눅스의 말씀으로 생각보다 빠르게 방향을 잡을 수 있었습니다.
  - 이전까지는 아마존 리눅스를 사용 중이었는데 mysql을 사용하려면 우분투로 인스턴스를 재생성하는 것이 낫다고 말씀해 주셔서 처음부터 우분투로 변경하는 시간을 가졌습니다.
- 리눅스보다 우분투가 메모리를 더 많이 잡아먹는 것인지는 몰라도 mysql만 깔면 메모리가 꽉 차서 더 이상 작동하지 않는 버그가 발생했습니다.
  - 이는 스왑 메모리 설정으로 해결할 수 있었습니다.
  - 외에도 컴파일 할 때 무한로딩 되는 버그가 있었는데 구글링으로 해결할 수 있었습니다.
- `build.gradle` 파일과 `application.preperties`를 수정하여 연동에 성공했습니다.
  - `application.preperties`는 깃허브 삭제 후 vim으로 우분투에서 직접 생성해 주었습니다. (보안 문제)

<br>

## soft-delete

- 게시물과 댓글을 삭제할 때, 칼럼 자체를 삭제했었습니다.
-  deleted라는 `BOOLEAN` 변수를 추가하여 `TRUE`일 때 `SELECT` 해오는 방식으로 재구현했습니다.
- 아래는 `Reply`와 `Article`의 레포지토리 구현 메서드입니다.

<br>

```java
    @Override
    public void deleteReply(Long id) {
        String articleDeletedSql = "update REPLIES set deleted=true where id=:id";

        Map<String, Object> param = Map.of("id", id);
        template.update(articleDeletedSql, param);
    }
```

```java
    @Override
    public void deleteArticle(Long id) {
        String articleDeletedSql = "update ARTICLES set deleted=true where id=:id";

        Map<String, Object> param = Map.of("id", id);
        template.update(articleDeletedSql, param);
    }
```

<br>

## 댓글 저장 리턴값 수정

- ReplyContoller의 `postReply()` 메서드
- ReplyService의 `saveReply()` 메서드
- JdbcTemplateReplyRepository의 `save()` 메서드

<br>

- ajax를 위해 `@RestController`로 바꾸던 도중 위 메서드들의 리턴값을 수정해야 했습니다.
- 원래 동기적 댓글 작성과 삭제를 할 때는 저장을 해줄 때 id 값을 리턴받을 필요가 없었습니다.
  - 왜냐하면 ajax와는 달리 댓글 작성을 누르면 새로고침이 되어 이미 데이터베이스가 저장되어 있으니 `PathVariable`에서 빼오면 그만이기 때문입니다.
  - id 값도 같이 리턴하기 위해 위 메서드를 수정한 것입니다.

<br>

```java
    @Override
    public Long save(Reply reply) {
        String sql = "insert into REPLIES (article_id, user_id, reply_content, reply_time, deleted) " +
                "values (:articleId, :userId, :replyContent, :replyTime, false)";
        KeyHolder keyHolder = new GeneratedKeyHolder();
        SqlParameterSource param = new BeanPropertySqlParameterSource(reply);
        template.update(sql, param, keyHolder);
        return (Objects.requireNonNull(keyHolder.getKey())).longValue();
    }
```

- 레포지토리에 `KeyHolder`를 추가하여 Long형 id 값을 리턴하도록 변경했습니다.
- 이 과정에서 ReplyRepository의 리턴 타입을 Long 타입으로 변경하였습니다.

<br>

```java
    public ReplyTimeForm saveReply(Long articleId, ReplyForm form, String userId) {
        Reply reply = new Reply(articleId, userId, form.getReplyContent());
        Long replyId = replyRepository.save(reply);
        return ReplyTimeForm.from(replyId, reply);
    }
```

```java
    public static ReplyTimeForm from(Reply reply) {
        return new ReplyTimeForm(reply.getId(), reply.getArticleId(), reply.getUserId(), reply.getReplyContent(), reply.getReplyTime());
    }

    public static ReplyTimeForm from(Long replyId, Reply reply) {
        return new ReplyTimeForm(replyId, reply.getArticleId(), reply.getUserId(), reply.getReplyContent(), reply.getReplyTime());
    }
```

- 서비스에서 ReplyTimeForm(Dto)를 이용하여 기존 Reply 객체와, 가져온 replyId를 리턴하게끔 수정했습니다.
- 이 과정에서 두번째 코드처럼 `from()` 메서드를 오버로딩 하였습니다.

<br>

```java
    @PostMapping("/questions/{articleId}/replies")
    public ReplyTimeForm postReply(@PathVariable Long articleId, ReplyForm form, HttpSession session) {
        return replyService.saveReply(articleId, form, (String) session.getAttribute(SessionConst.LOGIN_USER_ID));
    }
```

- 이전까지의 문제는 이 리턴값에 id가 null 값으로 들어왔다는 점인데, 이렇게 변경해준 이후로는 로그를 찍어봤을 때 원하는 결과가 도출됩니다.

<br>

## count(*)로 칼럼 카운트

- 아래는 각각 Article 레포지토리의 `findById()`와 `findAll()` 메서드입니다.

```java
    @Override
    public Optional<Article> findById(Long id) {
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id, " +
                "(select count(*) from REPLIES r where a.id=r.article_id and r.deleted=false) as replyCount " +
                "from ARTICLES a join USERS u on a.user_id=u.user_id " +
                "where a.id=:id and a.deleted=false";

        try {
            Map<String, Object> param = Map.of("id", id);
            Article article = template.queryForObject(sql, param, articleRowMapper);
            return Optional.ofNullable(article);
        } catch (EmptyResultDataAccessException e) {
            return Optional.empty();
        }
    }
```

<br>

```java
    @Override
    public List<Article> findAll() {
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id, " +
                "(select count(*) from REPLIES r where a.id=r.article_id and r.deleted=false) as replyCount " +
                "from ARTICLES a join USERS u on a.user_id=u.user_id " +
                "where a.deleted=false order by a.id desc";

        return template.query(sql, articleRowMapper);
    }
```

<br>

- 댓글은 여러 테이블의 조회가 필요하니 `join`을 이용하여 구현했습니다.
- 이 과정에서 댓글의 수를 가지고 있는 replyCount도 db 변수에 추가를 해야 하나 고민을 많이 했었습니다.
  - 그러던 중 파파베어의 마스터클래스를 듣고 `count(*)`의 존재를 알 수 있었습니다.
- REPLIES 테이블에서 조건에 따라 칼럼 count를 해주고 그것을 replyCount라는 변수에 추가하도록 했습니다.
  - 이 replyCount는 Entity(Reply)와 Dto(ReplyTimeForm)에서 처리해 줍니다.

<br>

# 해결하지 못한 부분

- **동료의 도움으로 해결했습니다 :)**

- ~~ajax 댓글 작성을 구현하면서 이틀을 통째로 날렸던 것 같습니다.~~
  - ~~정말 사소한 실수들로 고쳐진 문제들이 한둘이 아니어서 해결한 부분은 제외하고, 현재 가지고 있는 문제만 말씀드리겠습니다.~~

<br>

![ajax 버그](https://user-images.githubusercontent.com/62871026/233579381-a9452d4f-3232-4c8e-8a52-474343336af5.gif)

- ~~댓글을 입력하고, Network에 reply로는 잘 나타나고 있습니다.~~
- ~~유저 아이디, 댓글 입력 시간, 댓글 내용도 잘 가져오고 있습니다.~~

<br>
 
-  ~~**하지만 `답변하기` 버튼을 눌렀을 때, 댓글 형식이 아닌 폼 아래에 계속 추가되는 문제**가 발생합니다.~~
  - ~~새로고침을 해주면 해결이 되지만, 이러면 ajax로 처리해준 의미가 없다고 생각합니다.~~
- ~~스쿼드 세션 중 한 동료의 말씀으로 `qna/show.html`에서 경계를 나누면 된다는 것을 알 수 있었습니다.~~
  - ~~하지만 그 경계를 나눈다는 것이 정확한 어떠한 의미를 내포하는지는 아직까지도 이해하지 못했습니다.~~
- ~~아래의 두 코드만 보면 큰 문제는 없는 것 같은데 이런 문제를 해결하려면 어떻게 해야 할까요?~~
  - ~~이게 해결이 된 후 삭제 기능을 구현하고 싶은데 한 문제를 해결하면 또 다른 문제가 발생해서 넘어가질 못 하고 있습니다 🥲~~

<br>

```html
<form class="submit-write">

    <!--                            <form class="submit-write" method="POST" action="#"-->

    <!--                                  th:action="@{/questions/{articleId}/replies(articleId=${article.id})}">-->

    <div class="form-group" style="padding:14px;">

        <textarea name="replyContent" class="form-control"

                  placeholder="Update your status"></textarea>

    </div>

    <!--                                TODO: 경계 나누기 (입력 폼 아래로 출력되는 버그)-->

    <button id="post-reply" class="btn btn-success pull-right" type="button">답변하기</button>

    <div class="clearfix"/>

</form>

```

- ~~위는 `templates/qna/show.html`의 코드 중 일부입니다.~~

<br>

```javascript
$('#post-reply').click(() => {
    var queryString = $(".submit-write textarea[name=replyContent]").serialize();
    var url = window.location.pathname + '/replies';
    var articleId = window.location.pathname.split('questions/')[1];

    $.ajax({
        type: 'POST',
        url: url,
        data: queryString,
        dataType: 'json',
    }).done((db) => {
        console.log(db);
        var answerTemplate = $("#answerTemplate").html();
        var template = answerTemplate.format(db.userId, db.replyTime, db.replyContent, articleId, db.id);

        $(".qna-comment-slipp-articles").append(template);
        $("textarea[name=replyContent]").val("");
    }).fail((err) => {
        alert(JSON.stringify(err));
    });
});
```

- ~~위는 `static/js/scripts.js`의 코드 중 일부입니다.~~

<br>

# 버그 해결 완료

![Apr-21-2023 18-48-25](https://user-images.githubusercontent.com/62871026/233606166-7617e98b-b9ed-46ed-8bdc-0dcd25e9a259.gif)

- [이 커밋](https://github.com/won4885/be-java-cafe-max/commit/3bc14e0db4e4165b974d8fd06c9d574559234c9a)으로 해결하였습니다 :)

Commit:
bug: 글 수정할 때도 타임리프 에러

Commit:
fix(roach 피드백): Session 자체를 넘기지 말고 그 정보만을 넘기도록 수정

Commit:
fix(roach 피드백): 패스워드 연산 캡슐화

객체지향에서는 클래스가 가지고 있는 데이터를 통한 로직(연산) 들을 수행할때 클래스에게 데이터에 대한 책임을 주는 것이 보통 기본이기 때문이다.

Commit:
fix(roach 피드백): PutMapping 오류 드디어 해결

Model을 PutMapping에도 보내줘야 함

Commit:
fix: password -> existingPassword

field를 password로 2개를 써서 existingPassword에 null 값이 들어왔던 거였음

Commit:
fix: th:errors -> th:field

이것도 정말 단순하게 field 자리에 error를 넣어서 생긴 문제
정신 차려!!

Commit:
feat: MySql 8 연동 설정

Commit:
chore(테이블): 소문자는 인식을 못 하여 대문자로 수정

Commit:
chore: sql path 추가

Commit:
feat: allowPublicKeyRetrieval 설정 추가

Commit:
delete: mysql 사용으로 필요 없어진 sql 파일 삭제

Commit:
docs: application.properties -> git ignore 추가

Commit:
Fixed untracked files

Commit:
feat: Reply Entity 생성

댓글의 정보를 다 담고 있는 엔티티

Commit:
feat: 댓글 내용만을 담고 있는 Reply Dto 생성

Commit:
feat: LocalDateTime -> String Format 변환하는 Dto 생성

Commit:
feat: 댓글 기능을 담고 있는 레포지토리 인터페이스 생성

Commit:
feat: 댓글 기능을 담고 있는 레포지토리 구현

정보 저장 기능, 고유 Id에 따라 데이터 찾는 기능, 고유 Id와 articleId를 Join으로 찾는 기능

Commit:
feat: 댓글 레포지토리 객체를 가지고 변환하는 서비스 구현

Commit:
feat: 댓글 달면 그 정보를 post 방식으로 보내게끔 PostMapping 구현

Commit:
feat: Article과 Reply 정보를 둘 다 model로 보내주도록 추가

Commit:
feat: Model로 받아온 replies를 각 조건에 맞게 출력

Commit:
chore: 오타 수정

Commit:
refactor(pie): 페어 리뷰 때 깨달은 부분 리팩토링

로직 자체는 바뀌지 않음

Commit:
refactor: DB 컬럼명 변경

DB는 카멜이 안 좋은 거 같음

Commit:
fix: current_time -> currentTime

변수명 변경. current_time으로 하면 어디서 오류남

Commit:
fix: 댓글 view 성공

on (어쩌구)
이게 문제였다.

Commit:
refactor: soft delete 적용

그 과정에서 SimpleJdbcTemplate 제거

Commit:
feat: soft-delete 방식인 댓글 삭제 구현

validate와 return 경로의 의문점이 남아있다.

Commit:
fix: articleId 들어오도록 수정

1. repository에 articleId를 select하도록 추가해줌
2. 타임리프로는 왜 안 되는지는 모르겠지만 html에 input 태그를 추가하여 재구현

Commit:
feat: 자신이 작성한 댓글만 삭제 가능하도록 구현

Commit:
feat: count(*)를 이용하여 댓글 수 카운트

Commit:
feat: 게시글 id로 정렬되게 구현

Commit:
feat: replyCount 변수 추가

Commit:
feat: replyCount 출력

Commit:
fix: script 경로 버그 수정

Commit:
refactor: id를 리턴하도록 하여 id 값을 받아오도록 구현

ajax의 삭제 기능을 위해.
id가 null이면 못 불러오기 때문

Commit:
feat: id 값을 ReplyTimeForm(Dto)에 추가

그 과정에서 from 메서드를 오버로딩을 하게 됨

Commit:
feat: RestController로 변경 후 Dto 리턴

Commit:
fix: 스크립트 요구사항에 맞도록 수정

아직 입력 폼 아래로 출력되는 버그가 존재

Commit:
feat: ajax 기능 구현

Commit:
delete: 불필요한 코드 제거

Commit:
comment: 주석 수정

Commit:
fix: ajax 댓글 입력 버그 수정

form을 바깥으로 빼주면 됐음. 피아 = 신

Commit:
comment: TODO 주석 삭제

tmdgusya commented on src/main/java/kr/codesqaud/cafe/domain/dto/reply/ReplyTimeForm.java:
```diff
@@ -0,0 +1,73 @@
+        return new ReplyTimeForm(reply.getId(), reply.getArticleId(), reply.getUserId(), reply.getReplyContent(), reply.getReplyTime());
+    }
+
+    public static ReplyTimeForm from(Long replyId, Reply reply) {
```
문맥을 전달하는데 큰 차이는 없으나, 보통 단일 인자일경우 `from` 을 많이 사용하고요. 두개 이상의 인자인 경우 `of` 를 많이 사용합니다.

devbattery replied:
감사합니다! 수정하였습니다 :)

- [수정 커밋](https://github.com/won4885/be-java-cafe-max/commit/0b2b43366b0ddbc02dacdb5b1026512e0811a469)

tmdgusya commented on src/main/java/kr/codesqaud/cafe/repository/article/JdbcTemplateArticleRepository.java:
```diff
@@ -19,27 +18,27 @@
 
     @Override
     public Article save(Article article) {
+        String sql = "insert into ARTICLES (user_id, title, contents, currentTime, deleted) " +
```
테이블 명은 한번에 많은 부분을 바뀌여야 하는 상수이니 `ARTICLES` 은 상수로 관리하는게 괜찮아 보입니다.

won4885 replied:
감사합니다! 앞으로 테이블명은 상수로 관리해 주겠습니다.
또한 여러 레포지토리에서 중복 사용이 있으므로 Enum을 사용하여 변경해 보았습니다 🙇🏻

- [수정 커밋]https://github.com/won4885/be-java-cafe-max/commit/5bb80ee0c09eb70df2e65a0d9038ae00e6f293b6()

devbattery replied:
감사합니다! 앞으로 테이블명은 상수로 관리해 주겠습니다.
또한 여러 레포지토리에서 중복 사용이 있으므로 Enum을 사용하여 변경해 보았습니다 🙇🏻

- [수정 커밋](https://github.com/won4885/be-java-cafe-max/commit/5bb80ee0c09eb70df2e65a0d9038ae00e6f293b6)

tmdgusya commented on src/main/java/kr/codesqaud/cafe/domain/Reply.java:
```diff
@@ -0,0 +1,61 @@
+    private String replyContent;
+    private LocalDateTime replyTime;
+
+    public Reply() {
+    }
```
기본 생성자를 만드신 이유가 무엇일까요?

tmdgusya replied:
JDBC 등 Template 때문이라면 혹시 `public` 이 아닌 다른 접근 제어자로도 가능한지 한번 찾아보시길 바랍니다.

devbattery replied:
자바의 정석에서 기본 생성자는 항상 명시해주는 것이 좋다고 배운 뒤로는 계속 생성해 주고 있습니다.
말씀해주신 대로 그냥 명시하는 목적으로만 만들었기 때문에 private으로 바꿔주는 것이 좋을 것 같습니다!

- [수정 커밋](https://github.com/won4885/be-java-cafe-max/commit/a86126a7fb23bd90acba16767be3683a66401527)

tmdgusya reviewed (approved):
잘 작성해주셨네요 👍 

일단 올려주신 글중에서 생각해볼 점이 하나 있습니다.

```
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id, " +
                "(select count(*) from REPLIES r where a.id=r.article_id and r.deleted=false) as replyCount " +
                "from ARTICLES a join USERS u on a.user_id=u.user_id " +
                "where a.id=:id and a.deleted=false";
```


위와 같이 쿼리를 작성하게 되면 한가지 Trade-Off 가 생기게 되는데요. Article 을 가져올때마다 count 를 계산해야 해서 DB 에 부하를 줄 수 있습니다. 따라서 설리가 적어주신대로 ReplyCount 자체를 저장하는 방법도 있습니다 ㅎ. 그렇게 되면 어플리케이션 로직에서 동시성 문제를 겪게될수도 있는데요. 예를 들면, 자신은 replyCount 를 올렸다고 생각하지만 안올라가는 경우라던가, 뭐가 좋고 나쁜지는 상황에 따라 다른데 한번 무언가 쿼리를 작성하시가나 테이블관련작업을 하실때는 Trade-off 를 생각해보시는 것도 좋습니다.

won4885 commented:
> 위와 같이 쿼리를 작성하게 되면 한가지 Trade-Off 가 생기게 되는데요. Article 을 가져올때마다 count 를 계산해야 해서 DB 에 부하를 줄 수 있습니다. 따라서 설리가 적어주신대로 ReplyCount 자체를 저장하는 방법도 있습니다 ㅎ. 그렇게 되면 어플리케이션 로직에서 동시성 문제를 겪게될수도 있는데요. 예를 들면, 자신은 replyCount 를 올렸다고 생각하지만 안올라가는 경우라던가, 뭐가 좋고 나쁜지는 상황에 따라 다른데 한번 무언가 쿼리를 작성하시가나 테이블관련작업을 하실때는 Trade-off 를 생각해보시는 것도 좋습니다.

감사합니다! 말씀해주신 부분을 수정해 보았습니다 🙇🏻

<br>

1. 동시성 문제가 발생할 수 있는 경우를 해결하기 위해  `synchronized` 키워드를 추가해 주었습니다.
  - 하지만 이 키워드는 성능에 영향을 줄 수 있다는 점을 유의하려고 합니다.

```java
public synchronized Optional<Article> findById(Long id)
```

```java
public synchronized List<Article> findAll()
```

<br>

2. Article을 가져올 때마다 count를 계산해야 하는 문제를 수정했습니다.
  - ARTICLES 전체를 카운트하는 것이 아니라 `article_id`만 카운트하도록 했습니다.
  - 각각 `findById()`와 `findAll()`의 순서입니다.

```java
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, " +
                "count(r.article_id) as replyCount " +
                "from " + DataType.ARTICLES.getType() + " a " +
                "left join " + DataType.REPLIES.getType() + " r " + "on a.id=r.article_id and r.deleted=false " +
                "where a.id=:id and a.deleted=false " +
                "group by a.id, a.title, a.contents, a.currentTime, a.user_id";
```

```java
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id, " +
                "count(r.article_id) as replyCount " +
                "from " + DataType.ARTICLES.getType() + " a " +
                "join " + DataType.USERS.getType() + " u on a.user_id = u.user_id " +
                "left join " + DataType.REPLIES.getType() + " r on a.id = r.article_id and r.deleted = false " +
                "where a.deleted = false " +
                "group by a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id " +
                "order by a.id desc";
```

<br>

- [수정 커밋]()

devbattery commented:
@tmdgusya 

> 위와 같이 쿼리를 작성하게 되면 한가지 Trade-Off 가 생기게 되는데요. Article 을 가져올때마다 count 를 계산해야 해서 DB 에 부하를 줄 수 있습니다. 따라서 설리가 적어주신대로 ReplyCount 자체를 저장하는 방법도 있습니다 ㅎ. 그렇게 되면 어플리케이션 로직에서 동시성 문제를 겪게될수도 있는데요. 예를 들면, 자신은 replyCount 를 올렸다고 생각하지만 안올라가는 경우라던가, 뭐가 좋고 나쁜지는 상황에 따라 다른데 한번 무언가 쿼리를 작성하시가나 테이블관련작업을 하실때는 Trade-off 를 생각해보시는 것도 좋습니다.

감사합니다! 말씀해주신 부분을 수정해 보았습니다 🙇🏻

<br>

1. 동시성 문제가 발생할 수 있는 경우를 해결하기 위해  `synchronized` 키워드를 추가해 주었습니다.
  - 하지만 이 키워드는 성능에 영향을 줄 수 있다는 점을 유의하려고 합니다.

```java
public synchronized Optional<Article> findById(Long id)
```

```java
public synchronized List<Article> findAll()
```

<br>

2. Article을 가져올 때마다 count를 계산해야 하는 문제를 수정했습니다.
  - ARTICLES 전체를 카운트하는 것이 아니라 `article_id`만 카운트하도록 했습니다.
    - 이를 위해 `group by`를 사용해 주었습니다.
  - 각각 `findById()`와 `findAll()`의 순서입니다.

```java
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, " +
                "count(r.article_id) as replyCount " +
                "from " + DataType.ARTICLES.getType() + " a " +
                "left join " + DataType.REPLIES.getType() + " r " + "on a.id=r.article_id and r.deleted=false " +
                "where a.id=:id and a.deleted=false " +
                "group by a.id, a.title, a.contents, a.currentTime, a.user_id";
```

```java
        String sql = "select a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id, " +
                "count(r.article_id) as replyCount " +
                "from " + DataType.ARTICLES.getType() + " a " +
                "join " + DataType.USERS.getType() + " u on a.user_id = u.user_id " +
                "left join " + DataType.REPLIES.getType() + " r on a.id = r.article_id and r.deleted = false " +
                "where a.deleted = false " +
                "group by a.id, a.title, a.contents, a.currentTime, a.user_id, u.user_id " +
                "order by a.id desc";
```

<br>

- [수정 커밋](https://github.com/won4885/be-java-cafe-max/commit/c973a7037fe6bd110e99b66411ef1fd89097e76c)
